File: src/Ccmc.Core/Services/McpConfigReader.cs
using System.Text.Json;
using Ccmc.Core.Models;

namespace Ccmc.Core.Services;

/// <summary>
/// Reads the project's .mcp.json "mcpServers" map (read-only). Defensive: any malformed shape
/// yields an empty list rather than throwing — MCP config schema may evolve across CLI releases.
/// </summary>
public static class McpConfigReader
{
    public static IReadOnlyList<McpServerInfo> Read(string projectPath)
    {
        var result = new List<McpServerInfo>();
        try
        {
            var path = Path.Combine(projectPath, ".mcp.json");
            if (!File.Exists(path)) return result;
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            if (doc.RootElement.ValueKind != JsonValueKind.Object ||
                !doc.RootElement.TryGetProperty("mcpServers", out var servers) ||
                servers.ValueKind != JsonValueKind.Object)
                return result;

            foreach (var entry in servers.EnumerateObject())
            {
                var transport = "stdio";
                if (entry.Value.ValueKind == JsonValueKind.Object)
                {
                    if (entry.Value.TryGetProperty("type", out var t) && t.ValueKind == JsonValueKind.String)
                        transport = t.GetString() ?? "stdio";
                    else if (entry.Value.TryGetProperty("command", out var c) && c.ValueKind == JsonValueKind.String)
                        transport = c.GetString() ?? "stdio";
                }
                result.Add(new McpServerInfo(entry.Name, transport));
            }
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException or JsonException) { }
        return result;
    }

    public static bool Has(string projectPath) => Read(projectPath).Count > 0;
}

File: electron/src/core/config/mcpConfigReader.ts
/**
 * Pure parse logic for .mcp.json "mcpServers" maps.
 * Takes a JSON string → McpServerInfo[]. File reading is the caller's responsibility.
 *
 * Defensive: any malformed shape yields an empty list rather than throwing —
 * MCP config schema may evolve across CLI releases.
 *
 * Direct port of C# McpConfigReader (fs reads deferred to Phase 3 main-process wrapper).
 */

import type { McpServerInfo } from '../models'

/**
 * Parse a .mcp.json content string and return the list of MCP servers.
 * Pass null to represent an absent file — returns empty list.
 *
 * Transport resolution (mirrors C# logic):
 *   - If entry has a "type" string field → transport = that value
 *   - Else if entry has a "command" string field → transport = that value
 *   - Else → transport = "stdio"
 */
export function parseJson(json: string | null): McpServerInfo[] {
  if (json === null) return []
  const result: McpServerInfo[] = []
  try {
    const root: unknown = JSON.parse(json)
    if (typeof root !== 'object' || root === null || Array.isArray(root)) return result
    const servers = (root as Record<string, unknown>)['mcpServers']
    if (typeof servers !== 'object' || servers === null || Array.isArray(servers)) return result

    for (const [name, entry] of Object.entries(servers as Record<string, unknown>)) {
      let transport = 'stdio'
      if (typeof entry === 'object' && entry !== null && !Array.isArray(entry)) {
        const e = entry as Record<string, unknown>
        if (typeof e['type'] === 'string') {
          transport = e['type'] as string
        } else if (typeof e['command'] === 'string') {
          transport = e['command'] as string
        }
      }
      result.push({ name, transport })
    }
  } catch {
    // malformed JSON — return empty list, never throw
  }
  return result
}

/** Returns true when the JSON contains at least one MCP server. */
export function has(json: string | null): boolean {
  return parseJson(json).length > 0
}
